### PR TITLE
WIP log_backend_memory

### PR DESCRIPTION
**Pull request by laurenmurphyx64**
A really rough WIP of a memory logging backend. I'm still in the process of making changes to match with feedback I got recently, so the code is more up as a skeleton of what I'm doing than a prototype. For those wondering what the memory backend would do, here's the rundown:

> The memory backend writes logs to a buffer in memory, either in RAM or flash (or disk, in the future?). It can cache characters passed through `char_out` into a block buffer and write blocks at a time into the memory buffer to avoid flash wear. Using RAM allows for a host system to read logs live or post-crash if it has access to RAM. Using flash allows logs to be read post-mortem.

You can see the Kconfig options in `Kconfig.backends` for more information.

This backend would also have general uses, but I'm introducing it specifically to support flushing logs to memory on crash:

> If the LOG_FLUSH_ON_CRASH option is selected, the log_memory_backend will be selected but remain deactivated until kernel panic. When it receives the logging panic notification, it will activate and begin processing log messages.

Feedback would still be appreciated on the Kconfig options and (the many, possibly unnecessary) BUILD_ASSERTS. I also have some other questions:

- How can I let users plug in a flash partition to store the memory buffer - Kconfig, Devicetree, linker script fragment...?
- The winstream API isn't designed to work with flash, so unless I re-implement or modify it, users who want to use flash can't enable the PRESERVE_NEWEST option if the memory buffer overflows
- Should I implement the memory backend as separate backends (flash, memory/RAM, disk...) instead of one?

### CODE REVIEW

File: samples/hello_world/prj.conf [hunk 1]
```diff
@@ -1 +1,11 @@
-# nothing here
```
Comment (nashif) [inline samples/hello_world/prj.conf:1]: do not use hello_world for testing

File: subsys/logging/log_backend_memory.c [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	uint32_t flag = IS_ENABLED(CONFIG_LOG_BACKEND_MEMORY_SYST_ENABLE) ?
+		LOG_OUTPUT_FLAG_FORMAT_SYST : 0;
+
+	log_backend_std_sync_hexdump(&log_output_memory, flag, src_level,
+				     timestamp, metadata, data, length);
+}
+
+const struct log_backend_api log_backend_memory_api = {
+	.process = IS_ENABLED(CONFIG_LOG2) ? process : NULL,
+	.put = IS_ENABLED(CONFIG_LOG1_DEFERRED) ? put : NULL,
```
Comment (nashif) [inline subsys/logging/log_backend_memory.c:201]: please make this work for v2 only, no need to support v1 which is deprecated.

File: subsys/logging/Kconfig.backends [hunk 3]
```diff
@@ -380,4 +380,94 @@ config LOG_BACKEND_FS_FILES_LIMIT
 
 endif # LOG_BACKEND_FS
 
+config LOG_BACKEND_MEMORY
```
Comment (nashif) [inline subsys/logging/Kconfig.backends:383]: it is time to start having single Kconfig files for each backend I guess...

**Review by nashif — CHANGES_REQUESTED** [review]
first pass

File: subsys/logging/Kconfig.backends [hunk 4]
```diff
@@ -380,4 +380,94 @@ config LOG_BACKEND_FS_FILES_LIMIT
...
+config LOG_BACKEND_MEMORY_BUFFER_SIZE
+	int "Size of memory buffer in bytes"
+	help
+	  Must be a multiple of LOG_BACKEND_MEMORY_BLOCK_SIZE. If 
+	  LOG_BACKEND_MEMORY_PRESERVE_NEWEST is enabled, must also be big enough
+	  to hold the sys_winstream struct.
+
+config LOG_BACKEND_MEMORY_BLOCK_BUFFER_SIZE
+	int "Size of block in bytes to write to memory at a time"
+	# TODO: Are these reasonable defaults? First is a sentence, second is paragraph
```
Comment (teburd) [inline subsys/logging/Kconfig.backends:440]: This is fine, err on the side of larger is better here, when people run out of memory they will start tweaking the knobs. If you can capture a typical crash dump its fine.

**Review by teburd — COMMENTED** [review]
This is mostly ok, I'd work towards creating a dedicated test case or sample for this rather than using hello world for testing. This helps better understand how its meant to be configured and use in a sandbox type application.

**Pull request closed without merging**